File: PA1/cpu.py
#!/usr/bin/python

import thread
import time
import threading
import sys, getopt
import math

#Number of operations that will run on each test
N_OPERATIONS = 10**8
#Number of tests for each operation type
N_TESTS = 3
#Execution time for each test
TIME_INT = []
TIME_FL = []

def add_integer( thread, n ):
	count = 0
	startTime = time.clock()
	for i in xrange(n):	
		count += 1
	totalTime = time.clock() - startTime
	#print "%s: %s" % (thread, totalTime)

def add_float( thread, n ):
        count = 0.00
	float_increment = 1.07
        startTime = time.clock()
        for i in xrange(n):
                count += float_increment
        totalTime = time.clock() - startTime
        #print "%s: %s" % (thread, totalTime)

def average( values ):
	sum_ = 0.00
	for val in range(1,N_TESTS+1):
                sum_ += values[-(val)]
        return sum_ /float(N_TESTS)

def stand_deviation( values, average ):
	variance = 0.00
	for val in range(1,N_TESTS+1):
                variance += ((average - values[-(val)])**2)
	variance = variance / N_TESTS
	return math.sqrt( variance )

#Calculates GIOPS or GFLOPS for a given performance time
def calc_gops( time):
	return N_OPERATIONS / (time*(10**9))

class myThread(threading.Thread):
	def __init__(self, threadID, name, opType, n):
		threading.Thread.__init__(self)
		self.threadID = threadID
		self.name = name
		self.opType = opType
		self.n = n
	def run(self):
		print "Started %s %s operations on %s" % (self.n,self.opType,self.name)
		if self.opType == "integer":
			add_integer(self.name, self.n)
		else:
			add_float(self.name, self.n)


def main(argv):
	numThreads = ''
	outputFile = ''
	try:
		opts, args = getopt.getopt(argv,"t:",[])
	except getopt.GetoptError:
		print "cpu.py -t <#threads>"
		sys.exit(2)
	for opt, arg in opts:
		if opt == '-t':
			if arg is None:
				print "cpu.py -t <#threads>"
				sys.exit(2)
			numThreads = arg
	
	print "\n\n========== CPU PERFORMANCE BECHMARKING ==========\n"
	n = int((N_OPERATIONS)/int(numThreads))

	#Testing time for an empty loop
        startTime = time.clock()
	for i in xrange(n):
                pass
        loopTime = time.clock() - startTime

	threads = []
	startTime = time.clock()
	for i in range(0, int(numThreads)):
		thread = myThread(i, "T"+str(i),"integer", n)
		thread.start()
		threads.append(thread)

	for t in threads:
		t.join()
	totalTime = time.clock() - startTime - loopTime
	TIME_INT.append(totalTime);
	giops = n / (totalTime*(10**9))
	print "   Total time : %s  ### GIOPS : %s" % (totalTime, giops)

	threads = []
        startTime = time.clock()
        for i in range(0, int(numThreads)):
                thread = myThread(i, "T"+str(i), "float", n)
                thread.start()
                threads.append(thread)

        for t in threads:
                t.join()
        totalTime = time.clock() - startTime - loopTime
	TIME_FL.append(totalTime)
        flops = n / (totalTime*(10**9))
        print "   Total time : %s  ### FLOPS : %s" % (totalTime, flops)


if __name__ == "__main__":
	if len(sys.argv) != 3:
		print 'Usage: cpu.py -t <number of thread>'
		sys.exit(1)
	for i in range(0,N_TESTS):
		print "----- Running test #%s" % (i)
		main(sys.argv[1:])
	
	sum_int = 0
	sum_float = 0
	avg_int = average(TIME_INT)
	avg_float = average(TIME_FL)
	sd_int = stand_deviation(TIME_INT, avg_int)
	sd_float = stand_deviation(TIME_FL, avg_float)
	
        print "\n======= SUMMARY OF CPU PERFORMANCE ======="
        print "Tests                                    : %s" % (N_TESTS)
	print "Number of threads                        : %s" % (sys.argv[2])
        print "Average time for integer operations      : %ss" % (avg_int)
        print "Standard deviation for integer operations: %ss" % (sd_int)
	print "Average integer operations per second    : %s GIOPS" %(calc_gops(avg_int)) 
        print "Average time for float operations        : %ss" % (avg_float)
        print "Standard deviation for float operations  : %ss" % (sd_float)
	print "Average float operations per second      : %s GFLOPS" %(calc_gops(avg_float)) 
        print "=====================================\n\n"

File: PA1/cpu.cpp
#include <time.h>
#include <stdio.h>
#include <math.h>
#include <pthread.h>
#include <iostream>

const int N = 1000;
const int N_TESTS = 1;
const float CONST_FLOAT = 0.7;
int time_int[] = {0, 0, 0};
int time_float[] = {0, 0, 0};

//Performs N^3 operations
void *integer_op(void *threadid){
	int A[N][N];

        //initialize matrix A
        for(int i=0;i<N;i++)
        {
                for(int j=0;j<N;j++)
                {
                        A[i][j] = 0;
                }
        }

	for(int i=0;i<N;i++)
        {
        	for(int j=0;j<N;j++)
                {
			for(int k = 0; k < N; k++)
			{
                        	A[i][j]++;
			}
                }
        }
	printf("\n  - End of a thread.\n");
	
}

//Performs N^3 operations
void *float_op(void *threadid){
	float A[N][N];

        //initialize matrix A
        for(int i=0;i<N;i++)
        {
                for(int j=0;j<N;j++)
                {
                        A[i][j] = 0.0;
                }
        }

	for(int i=0;i<N;i++)
        {
        	for(int j=0;j<N;j++)
                {
			for(int k = 0; k < N; k++)
			{
                        	A[i][j]+=CONST_FLOAT;
			}
                }
        }
	printf("\n  - End of a thread.\n");
	
}

void cpu_test(int n_threads, int test_number){
	pthread_t threads[n_threads];
	
	int rc;
	int i;
	long int total_time = 0;

	//INTEGER OPERATIONS
	clock_t start = clock(), diff;
    	for (i=0; i<n_threads; i++){
		rc = pthread_create(&threads[i], NULL, integer_op, (void *)i);
		printf("\nCREATED THREAD INTEGER%d\n", i);
    	}
	for (int i=0; i<n_threads; i++){
       		(void) pthread_join(threads[i], NULL);
    	}
	diff = clock() - start;
	time_int[test_number] = diff;

	//FLOAT OPERATIONS
	start = clock();
    	for (i=0; i<n_threads; i++){
		rc = pthread_create(&threads[i], NULL, float_op, (void *)i);
		printf("\nCREATED THREAD FLOAT%d\n", i);
    	}
	for (int i=0; i<n_threads; i++){
       		(void) pthread_join(threads[i], NULL);
    	}
	diff = clock() - start;
	time_float[test_number] = diff;
}


int main(int argc, char *argv[]) {
	if(argc != 2)
	{
		printf("\nUsage: <#thread>\n");
		return 1;
	}

	int n_threads = atoi(argv[1]);
	for(int i=0; i<N_TESTS; i++) {
		cpu_test(n_threads, i);
	}
	
	long int n_operations = pow(N,3) * n_threads;
	long int diff = time_int[0];
	double time = (double) diff / (double) CLOCKS_PER_SEC;
	printf("\nTime integer operations = %f seconds", time);
	printf("\nGIOPS = %f\n", n_operations/(time * pow(10,9)));

	diff = time_float[0];
	time = (double) diff / (double) CLOCKS_PER_SEC;
	printf("\nTime float operations = %f seconds", time);
	printf("\nFLOPS = %f\n", n_operations/(time * pow(10,9)));
}
